assets/pages/new-contributions-landing/components/PayPalRecurringButton.jsx

```diff
@@ -0,0 +1,81 @@
+// @flow
+
+// ----- Imports ----- //
+
+import ReactDOM from 'react-dom';
+import React from 'react';
+
+import type { Csrf as CsrfState } from 'helpers/csrf/csrfReducer';
+import { getPayPalOptions } from 'helpers/paymentIntegrations/newPaymentFlow/payPalRecurringCheckout';
+import type { IsoCurrency } from 'helpers/internationalisation/currency';
+import type { PaymentAuthorisation } from 'helpers/paymentIntegrations/newPaymentFlow/readerRevenueApis';
+import type { PayPalAuthorisation } from 'helpers/paymentIntegrations/newPaymentFlow/readerRevenueApis';
+
+
+type PropTypes = {|
+  onPaymentAuthorisation: PaymentAuthorisation => void,
+  csrf: CsrfState,
+  currencyId: IsoCurrency,
+  hasLoaded: boolean,
+  canOpen: () => boolean,
+  whenUnableToOpen: () => void,
+  formClassName: string,
+  isTestUser: boolean,
+  setupRecurringPayPalPayment: (Function, Function, IsoCurrency, CsrfState) => void,
+|};
+
+
+// Q. Why is this a class rather than a function?
+// A. Because we need to override shouldComponentUpdate.
+//
+// Q. Why do you need to override shouldComponentUpdate? Isn't that dangerous?
+// A. We want to bypass React's normal decision-making about whether to re-render
+// this particular component, because it uses some third-party code we can't control and
+// which behaves in a very specific way.
+//
+// When we use PayPal's JS library via window.paypal.Button.driver, it renders a button inside
+// an iframe. Each time this code is called, the iframe ends up with a different id,
+// so React will always try and update the DOM even when nothing has substantively changed.
+//
+// We don't want this to happen, for two reasons.
+// 1. Loading this iframe is an expensive operation which causes an obvious visual re-render
+// 2. We don't want to have to re-bind handlers which interact with the iframe
+//    (e.g. the handler bound in getPayPalOptions)
+export class PayPalRecurringButton extends React.Component<PropTypes> {
```

Comment:
could you extend React.PureComponent which is more explicit about not having state but gives you the lifecycle events that an actual function cannot

Reply:
Based on what I'm reading here the only difference is that React.Component doesn't implement `shouldComponentUpdate` but `React.PureComponent` implements it with a shallow prop and state comparison: https://reactjs.org/docs/react-api.html#reactpurecomponent.

But I'm not seeing stuff about it being explicit that it's stateless?

Reply:
yes sorry my misunderstanding totally, i was of thinking of React.SFC (but thats just a typescript type to represent a stateless functional component - i.e. can't extend it and thus can't override the `shouldComponentUpdate` lifecycle method)